src/metabase/search/scoring.clj

```diff
@@ -271,31 +275,53 @@
    [result]
    (weights-and-scores result))
 
+(defn- compute-normalized-score [scores]
+  (let [weight-sum (reduce + (map #(or (:weight %) 0) scores))
+        score-sum (reduce
+                   (fn [acc {:keys [weight score] :or {weight 0 score 0}}] (+ acc (* score weight)))
+                   0
+                   scores)]
+    (if (zero? weight-sum)
+      0
+      (/ score-sum weight-sum))))
+
+(defn force-weight
+  "Reweight `scores` such that the sum of their weights equals `total`, and their proportions do not change."
+  [scores total]
+  (let [total-found (reduce + (map :weight scores))]
+    (mapv #(update % :weight (fn [weight]
+                               (if (contains? #{nil 0} total-found)
+                                 0
+                                 (* total (/ weight total-found))))) scores)))
```

Comment:
Wouldn't it be clearer if we created a function before mapping? In the `(contains? #{nil 0} total-found)` case it'd be `(constantly 0)`.